state-chain/pallets/cf-swapping/src/lib.rs

```diff
@@ -1238,13 +1222,73 @@ pub mod pallet {
 			if violating.is_empty() {
 				Ok(non_violating)
 			} else {
-				Err(BatchExecutionError::PriceLimitHit {
-					successful_swaps: non_violating.into_iter().map(|ctx| ctx.swap).collect(),
-					failed_swaps: violating.into_iter().map(|ctx| ctx.swap).collect(),
+				Err(BatchExecutionError::PriceViolation {
+					violating_swaps: violating.into_iter().map(|ctx| ctx.swap).collect(),
+					non_violating_swaps: non_violating.into_iter().map(|ctx| ctx.swap).collect(),
 				})
 			}
 		}
 
+		/// Attempts to find (and execute) a batch of swaps that wouldn't result in hitting the
+		/// price impact limit, starting with the given batch, and taking swaps out of the batch if
+		/// needed.
+		#[transactional]
+		fn execute_batch(
+			mut swaps_to_execute: Vec<Swap<T>>,
+		) -> Result<BatchExecutionOutcomes<T>, DispatchError> {
+			let mut failed_swaps = vec![];
+
+			loop {
+				if swaps_to_execute.is_empty() {
+					return Ok(BatchExecutionOutcomes { successful_swaps: vec![], failed_swaps });
+				}
+
+				match Self::try_execute_without_violations(swaps_to_execute.clone()) {
+					Ok(successful_swaps) =>
+						return Ok(BatchExecutionOutcomes { successful_swaps, failed_swaps }),
+					Err(BatchExecutionError::SwapLegFailed {
+						asset,
+						direction,
+						amount,
+						failed_swap_group,
+					}) => {
+						Self::deposit_event(Event::<T>::BatchSwapFailed {
+							asset,
+							direction,
+							amount,
+						});
+
+						// Find the largest swap from the failing pool/direction:
+						let swap_to_remove = utilities::split_off_highest_impact_swap(
+							&mut swaps_to_execute,
+							&failed_swap_group,
+							direction,
+						);
+
+						// Abort if we couldn't find a swap to remove; otherwise, remove the
+						// swap and try again with the remaining swaps:
+						if let Some(swap) = swap_to_remove {
+							failed_swaps.push(swap);
+						} else {
+							failed_swaps.extend(swaps_to_execute);
+							return Ok(BatchExecutionOutcomes {
+								successful_swaps: vec![],
+								failed_swaps,
+							});
+						}
+					},
+					Err(BatchExecutionError::PriceViolation {
+						violating_swaps,
+						non_violating_swaps,
+					}) => {
+						failed_swaps.extend(violating_swaps);
+						swaps_to_execute = non_violating_swaps;
+					},
+					Err(BatchExecutionError::DispatchError { error }) => return Err(error),
```

Comment:
This is the only error condition in this loop, and it can only happen if we exceed the the transaction depth limit, right? So am I right in thinking that this function (execute_batch) doesn't need to be transactional (since it can never really error, and never needs to roll back)? 

ie. instead of passing this error back up the chain, we could just log it at this level.

Reply:
Right, I changed this to never return BatchExecutionOutcomes directly now.